<h1 style="text-align: center;">Desafío - Calculando probabilidades</h1>

- Para realizar este desafío debes haber estudiado previamente todo el material disponible correspondiente a la unidad.
- Una vez terminado el desafío, comprime la carpeta que contiene el desarrollo de los requerimientos solicitados y sube el .zip en el LMS.
- Desarrollo del desafío: individual
- Para la realización del desafío necesitarás apoyarte del archivo `Apoyo Desafío - Calculando probabilidades.zip`, disponible en la <a href="https://empieza.desafiolatam.com/cohorts/2108/sections/79626/activities/120911/items/617686">plataforma</a>.
- Puntaje total: 10 puntos.

## Habilidades a evaluar
- Utilizar conceptos básicos de probabilidad.

## Descripción
Un amigo, al cual le gusta mucho el fútbol, tiene en mente poder crear una app de apuestas para el próximo mundial. Para ello, solicita tu ayuda como analista de datos de manera de poder trabajar con las probabilidades de victoria o derrota de los equipos, y que así él diseñe el resto de la app. Tú aceptas y ordenas el trabajo que debes realizar a partir de los siguientes requerimientos.

## Requerimientos

### 1. Lectura de archivos (1 punto)
- Importa pandas y numpy siguiendo las convenciones.
- Lee el set de datos `worldcup2014.csv` y asígnela a una variable `df`.
- Solicita las primeras 5 observaciones con `head`.

In [1]:
# Utiliza esta celda para importar pandas y numpy
import pandas as pd
import numpy as np


In [2]:
# Utiliza esta celda para leer los datos y almacenarlos en la variable df
df = pd.read_csv('worldcup2014.csv')

In [3]:
# Utiliza esta celda para solicitar las primeras 5 observaciones con head
df.head()

team     continent group  group_pos  cantidad_juegos  \
0       Brazil  southamerica     A          1                3   
1       Mexico  northamerica     A          2                3   
2      Croatia        europe     A          3                3   
3     Cameroon        africa     A          4                3   
4  Netherlands        europe     B          1                3   

   juegos_ganados  juegos_empatados  juegos_perdidos  goles_favor  \
0               2                 1                0            7   
1               2                 1                0            4   
2               1                 0                2            6   
3               0                 0                3            1   
4               3                 0                0           10   

   goles_contra goles_diferencia  puntos  clasificado  
0             2                5       7            1  
1             1                3       7            1  
2             6                0       3            0  
3             9               −8       0            0  
4             3                7       9            1

In [4]:
#Se utiliza esta celda para explorar la cantidad de nulos para asegurar la calidad de los datos
df.isna().sum()

team                0
continent           0
group               0
group_pos           0
cantidad_juegos     0
juegos_ganados      0
juegos_empatados    0
juegos_perdidos     0
goles_favor         0
goles_contra        0
goles_diferencia    0
puntos              0
clasificado         0
dtype: int64

### 2. Frecuencia de `continent`: utilizando `value_counts` responde lo siguiente (1 punto)

In [18]:
# Utiliza esta celda para aplicar value_counts a la columna continent. Recordar que debemos mostrar frecuencias relativas para obtener las probabilidades.
df['continent'].value_counts()
df['continent'].value_counts(normalize=True) #Así se obtienen las frecuencias relativas al total de la muestra
df['continent'].value_counts('%') #Se puede realizar la misma operación que normalize agregando cualquier cadena de texto, ya que el primer parámetro de value_count es normalize y cualquier texto tiene el valor True

europe          0.40625
southamerica    0.18750
africa          0.15625
northamerica    0.12500
asia            0.12500
Name: continent, dtype: float64

**¿Cuál es el continente con una mayor presencia en la muestra?**

**Respuesta**: Europa

**¿Cuál es la probabilidad de elegir un equipo asiático al azar?**

**Respuesta:** 0.125 de 1 o 12.5%


**¿Cuál es la probabilidad de elegir un equipo africano al azar?**

**Respuesta:** 0.15625 de 1 o 15.625%

### 3. Generación de un subconjunto de datos: por cada continente, genere un nuevo objeto que almacene sólo las observaciones del continente (1 punto)
- _**Tip**_: Para ello puede utilizar la siguiente sintaxis: `df[df['variable'] == condicion]`.

In [6]:
# Se utiliza esta celda para almacenar todos los posibles continentes. Se puede lograr esto con unique()
continents = df['continent'].unique()
continents

array(['southamerica', 'northamerica', 'europe', 'africa', 'asia'],
      dtype=object)

In [20]:
# Se utiliza esta celda para generar una lista con los DataFrames que contienen los registros de cada contiennte
# df_europe = df[df["continent"] == "europe"]
# df_europe["continent"].unique()
lista_df = []
for c in continents:
    lista_df.append(df[df['continent']==c])

lista_df= [df[df['continent'] == c] for c in continents]
#Se muestra el DataFrame sólo con los datos de América del sur
lista_df[0]


team     continent group  group_pos  cantidad_juegos  juegos_ganados  \
0      Brazil  southamerica     A          1                3               2   
5       Chile  southamerica     B          2                3               2   
8    Colombia  southamerica     C          1                3               3   
13    Uruguay  southamerica     D          2                3               2   
18    Ecuador  southamerica     E          3                3               1   
20  Argentina  southamerica     F          1                3               3   

    juegos_empatados  juegos_perdidos  goles_favor  goles_contra  \
0                  1                0            7             2   
5                  0                1            5             3   
8                  0                0            9             2   
13                 0                1            4             4   
18                 1                1            3             3   
20                 0                0            6             3   

   goles_diferencia  puntos  clasificado  
0                 5       7            1  
5                 2       6            1  
8                 7       9            1  
13                0       6            1  
18                0       4            0  
20                3       9            1

In [22]:
lista_df[2] #Se muestra el DataFframe sólo con los datos de europa

team continent group  group_pos  cantidad_juegos  \
2                  Croatia    europe     A          3                3   
4              Netherlands    europe     B          1                3   
6                    Spain    europe     B          3                3   
9                   Greece    europe     C          2                3   
14                   Italy    europe     D          3                3   
15                 England    europe     D          4                3   
16                  France    europe     E          1                3   
17             Switzerland    europe     E          2                3   
22  Bosnia and Herzegovina    europe     F          3                3   
24                 Germany    europe     G          1                3   
26                Portugal    europe     G          3                3   
28                 Belgium    europe     H          1                3   
30                  Russia    europe     H          3                3   

    juegos_ganados  juegos_empatados  juegos_perdidos  goles_favor  \
2                1                 0                2            6   
4                3                 0                0           10   
6                1                 0                2            4   
9                1                 1                1            2   
14               1                 0                2            2   
15               0                 1                2            2   
16               2                 1                0            8   
17               2                 0                1            7   
22               1                 0                2            4   
24               2                 1                0            7   
26               1                 1                1            4   
28               3                 0                0            4   
30               0                 2                1            2   

    goles_contra goles_diferencia  puntos  clasificado  
2              6                0       3            0  
4              3                7       9            1  
6              7               −3       3            0  
9              4               −2       4            1  
14             3               −1       3            0  
15             4               −2       1            0  
16             2                6       7            1  
17             6                1       6            1  
22             4                0       3            0  
24             2                5       7            1  
26             7               −3       4            0  
28             1                3       9            1  
30             3               −1       2            0

### 4. Probabilidad de clasificación a la siguiente ronda: Calcula la probabilidad de que un país pase a la siguiente ronda (clasifique) por continente (4 puntos)
_**Tip**_: El problema puede ser resuelto de varias formas:
- La manera más intuitiva corresponde a calcular la probabilidad en cada subconjunto creado en el problema 3.
- Una forma alternativa es copiar todos los subsets creados en el problema 3 en una lista e iterar sobre dicha lista.
- Una forma más elegante consiste en utilizar una operación de agrupación por continente, y luego calcular la probabilidad.

In [23]:
# Utiliza esta celda para explorar las columnas del DataFrame e identificar la que indica si un país clasificó o no
df.columns

Index(['team', 'continent', 'group', 'group_pos', 'cantidad_juegos',
       'juegos_ganados', 'juegos_empatados', 'juegos_perdidos', 'goles_favor',
       'goles_contra', 'goles_diferencia', 'puntos', 'clasificado'],
      dtype='object')

In [24]:
df['clasificado'].value_counts()

1    16
0    16
Name: clasificado, dtype: int64

In [25]:
lista_df[0]['clasificado'].value_counts('%')

1    0.833333
0    0.166667
Name: clasificado, dtype: float64

In [31]:
# Se utiliza esta celda para observar la probabilidad de clasificación en cada uno de los DataFrame de los continentes creados en la pregunta anterior
# Recuerda que "1" indica "sí clasifica", y que de una Serie (como es el caso de lo entregado por value_counts) puedes acceder al valor asociado 
# a un índice con get

for df_continente in lista_df:
    probs_df = df_continente['clasificado'].value_counts('normalize')
    nombre_continente = df_continente['continent'].unique()
    print(f'La probabilidad de clasificar para un país del continente {nombre_continente[0]} es: {round(probs_df.get(1),3) if probs_df.get(1)!= None else 0}')

La probabilidad de clasificar para un país del continente southamerica es: 0.833
La probabilidad de clasificar para un país del continente northamerica es: 0.75
La probabilidad de clasificar para un país del continente europe es: 0.462
La probabilidad de clasificar para un país del continente africa es: 0.4
La probabilidad de clasificar para un país del continente asia es: 0


In [27]:
# Se puede lograr lo mismo usando el df general de todos los datos, usando groupby

df.groupby(['continent'])['clasificado'].value_counts('normalize')

continent     clasificado
africa        0              0.600000
              1              0.400000
asia          0              1.000000
europe        0              0.538462
              1              0.461538
northamerica  1              0.750000
              0              0.250000
southamerica  1              0.833333
              0              0.166667
Name: clasificado, dtype: float64

**¿Cuál es la probabilidad de que un país asiático pase a la siguiente ronda?**

**Respuesta:** 0

**¿Cuáles son los dos continentes con la mayor probabilidad de clasificar?**

**Respuesta:** Norteamérica y Suramérica

**¿Cuál es la probabilidad de que un país europeo no clasifique?**

**Respuesta:** 0.538462 de 1 o 53.85%

### 5. Probabilidad de victoria y derrota: Se pide utilizar `value_counts` para calcular la probabilidad de que un continente contenga países con o sin victorias (3 puntos).
- Para ello crea una columna adicional llamada `ha_ganado` que contenga un `0` si es que un país _no_ tiene juegos ganados, y un `1` si es que ha ganado al menos un juego.
- Sugerencias: Para crear dicha columna puede revisar:
    - `np.where`, de Numpy
    - `Series.where`, de Pandas

- Además puede utilizar los mismos procedimientos sugeridos en el Requerimiento 4.

In [32]:
# Se utiliza esta celda para resolver mediante np.where

df['ha_ganado'] = np.where(df['juegos_ganados'] > 0, 1, 0) 
df['ha_ganado'].value_counts()

1    23
0     9
Name: ha_ganado, dtype: int64

In [33]:
# Se utiliza esta celda para resolver mediante Series.where
df['ha_ganado'] = df['juegos_ganados'].where(lambda x: x == 0 ,1)
df['ha_ganado'].value_counts()

1    23
0     9
Name: ha_ganado, dtype: int64

In [34]:
# Se utiliza esta celda para mostrar las probabilidades usando groupby (con celda 'ha_ganado' ya creada)
df.groupby(['continent'])['ha_ganado'].value_counts('normalize')

continent     ha_ganado
africa        1            0.600000
              0            0.400000
asia          0            1.000000
europe        1            0.846154
              0            0.153846
northamerica  1            0.750000
              0            0.250000
southamerica  1            1.000000
Name: ha_ganado, dtype: float64

**A. ¿Qué continente tiene la mayor probabilidad de tener países con al menos 1 victoria?**

**Respuesta**: Suramérica

**B. ¿Qué continente tiene un nivel similar entre países que tienen o no tienen victorias?**

**Respuesta**: África

**Reformule el código para juegos perdidos y analice los continentes con mayor probabilidad de tener países sin juegos perdidos.**

In [43]:
# Utilice esta celda para generar la columna 'ha_perdido' indicando con 1 si el país ha perdido al menos 1 vez, y 0 si no

df['ha_perdido'] = np.where(df['juegos_perdidos'] > 0, 1, 0)
df['ha_perdido'].value_counts()

1    23
0     9
Name: ha_perdido, dtype: int64

In [44]:
# Se utiliza esta celda para mostrar las probabilidades por continente de haber perdido al menos 1 juego

df.groupby(df['continent'])['ha_perdido'].value_counts('%')

continent     ha_perdido
africa        1             1.000000
asia          1             1.000000
europe        1             0.692308
              0             0.307692
northamerica  0             0.500000
              1             0.500000
southamerica  0             0.500000
              1             0.500000
Name: ha_perdido, dtype: float64

**Respuesta:** Norteamérica y Suramérica

In [45]:
#Observación de algunos datos del DataFrame

df[['team','juegos_ganados','juegos_perdidos','ha_ganado','ha_perdido']]

team  juegos_ganados  juegos_perdidos  ha_ganado  \
0                   Brazil               2                0          1   
1                   Mexico               2                0          1   
2                  Croatia               1                2          1   
3                 Cameroon               0                3          0   
4              Netherlands               3                0          1   
5                    Chile               2                1          1   
6                    Spain               1                2          1   
7                Australia               0                3          0   
8                 Colombia               3                0          1   
9                   Greece               1                1          1   
10             Ivory Coast               1                2          1   
11                   Japan               0                2          0   
12              Costa Rica               2                0          1   
13                 Uruguay               2                1          1   
14                   Italy               1                2          1   
15                 England               0                2          0   
16                  France               2                0          1   
17             Switzerland               2                1          1   
18                 Ecuador               1                1          1   
19                Honduras               0                3          0   
20               Argentina               3                0          1   
21                 Nigeria               1                1          1   
22  Bosnia and Herzegovina               1                2          1   
23                    Iran               0                2          0   
24                 Germany               2                0          1   
25           United States               1                1          1   
26                Portugal               1                1          1   
27                   Ghana               0                2          0   
28                 Belgium               3                0          1   
29                 Algeria               1                1          1   
30                  Russia               0                1          0   
31             South Korea               0                2          0   

    ha_perdido  
0            0  
1            0  
2            1  
3            1  
4            0  
5            1  
6            1  
7            1  
8            0  
9            1  
10           1  
11           1  
12           0  
13           1  
14           1  
15           1  
16           0  
17           1  
18           1  
19           1  
20           0  
21           1  
22           1  
23           1  
24           0  
25           1  
26           1  
27           1  
28           0  
29           1  
30           1  
31           1